room/room-compiler-processing/src/main/java/androidx/room/compiler/processing/javac/JavacBasicAnnotationProcessor.kt

```diff
@@ -31,43 +34,58 @@ import javax.lang.model.element.Element
+    // the state is cleared at the end of each round in BasicAnnotationProcessor#postRound()
+    private var xEnv: JavacProcessingEnv? = null
+
     final override fun steps(): Iterable<Step> {
-        // Execute all processing steps in a single auto-common Step. This is done to share the
-        // XProcessingEnv and its cached across steps in the same round.
-        val steps = processingSteps()
-        val parentStep = object : Step {
-            override fun annotations() = steps.flatMap { it.annotations() }.toSet()
+        val delegatingSteps = processingSteps().map { DelegatingStep(it) }
+        val initializingStep = InitializingStep(delegatingSteps)
+        return mutableListOf<Step>(initializingStep) + delegatingSteps
+    }
+
+    /** A step that initializes the state before every processing round.  */
+    private inner class InitializingStep(val delegatingSteps: Iterable<DelegatingStep>): Step {
+        override fun annotations() = delegatingSteps.flatMap { it.annotations() }.toSet()
+
+        override fun process(
+            elementsByAnnotation: ImmutableSetMultimap<String, Element>
+        ): Set<Element> {
+            xEnv = JavacProcessingEnv(processingEnv)
+            return emptySet()
+        }
+    }
 
-            override fun process(
-                elementsByAnnotation: ImmutableSetMultimap<String, Element>
-            ): Set<Element> {
-                val xEnv = JavacProcessingEnv(processingEnv)
-                val convertedElementsByAnnotation = mutableMapOf<String, Set<XElement>>()
-                annotations().forEach { annotation ->
-                    convertedElementsByAnnotation[annotation] =
-                        elementsByAnnotation[annotation].mapNotNull { element ->
-                            xEnv.wrapAnnotatedElement(element, annotation)
-                        }.toSet()
-                }
-                val results = steps.flatMap { step ->
-                    step.process(
-                        env = xEnv,
-                        elementsByAnnotation = step.annotations().associateWith {
-                            convertedElementsByAnnotation[it] ?: emptySet()
-                        }
-                    )
-                }
-                return results.map { (it as JavacElement).element }.toSet()
+    /** A [Step] that delegates to an [XProcessingStep]. */
+    private inner class DelegatingStep(val xStep: XProcessingStep): Step {
+        override fun annotations() = xStep.annotations()
+
+        override fun process(
+            elementsByAnnotation: ImmutableSetMultimap<String, Element>
+        ): Set<Element> {
+            val xElementsByAnnotation = mutableMapOf<String, Set<XElement>>()
+            xStep.annotations().forEach { annotation ->
+                xElementsByAnnotation[annotation] =
+                    elementsByAnnotation[annotation].mapNotNull { element ->
+                        xEnv!!.wrapAnnotatedElement(element, annotation)
```

Comment:
Instead of using double bang here with `xEnv` you can create a local value to make sure environment is not null:
`val env = checkNotNull(xEnv)` and then use `env` instead of `xEnv!!`

Reply:
Done.